### PR TITLE
Fix legacy inpainting pipeline: scale mask_image tensor to latent size, and avoid adding noise to final latents

### PR DESCRIPTION
**Pull request by 1lint**
Hi Huggingface, thanks for making stable diffusion ML accessible for all (especially me 🤗). 
While making a stable diffusion webui, I came across some issues with the legacy inpainting pipeline for which I've made a minimal reproduction [here](https://github.com/1lint/sdpipe_webui/blob/2414bfde39df2fca0d1793a4e12e3c5bc8666ba3/inpaint_noise_issue.ipynb). The parent repository also contains a stable diffusion webui with inpainting support to reproduce the issues 2 & 3 below. 

1. When passing a tensor as the mask image, a shape mismatch error will be triggered because the pipeline does not resize the mask tensor to the latent shape, the pipeline only resizes the mask when it is a PIL image. The pipeline also needs to permute the tensor channel dimension to pytorch standard (B x C x H x W). Note the __call__ method documentation says the mask should have expected shape (B, H, W, 1) when it is a tensor. 
Fix: added logic to resize and permute mask dimensions when it is a tensor

2. Noise is still added to the final latents right before the latent decoding step, resulting in noisy image outputs. For high inference steps and some schedulers, the result may not be easily noticeable, but as seen in the linked notebook, the effect can be very problematic. I passed a mask tensor of all ones (i.e. do not repaint anything), yet there is significant corruption of the image due to the added noise right before decoding.
Fix: added line to restore un-noised original latents corresponding to unmasked portions of the original image right before the decoding step

3. Inpainting a subset of the latents can affect the entire decoded image, even portions of the image far away from the drawn mask. I've found this is a lot more noticeable for anime style diffusion models. 
Fix: after decoding inpainted image, restore unmasked portions of the original image, keeping only the parts of the inpainted image corresponding to the drawn mask

Thanks for taking a look and please let me know how I can improve the pull request. I've been cloning repos from github for a long time but this is my first pull request.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import PIL
 from packaging import version
 from transformers import CLIPFeatureExtractor, CLIPTextModel, CLIPTokenizer
+import torchvision
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:24]: `torchvision` is not a hard-dependency. We sadly cannot just import it here and should not use it in this script

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py [hunk 2]
```diff
@@ -508,7 +522,14 @@ def __call__(
...
         if not isinstance(mask_image, torch.FloatTensor):
-            mask_image = preprocess_mask(mask_image, self.vae_scale_factor)
+            mask_image, input_mask_image = preprocess_mask(mask_image, self.vae_scale_factor)
+        else:
+            # if mask_image is tensor, we still need to scale mask_image to the latent shape
+            # based on __call__ documentation, mask_image tensor is expected to be size (B, H, W, 1)
+            mask_image = mask_image.permute(0, 3, 1, 2) # permute mask to pytorch standard (B, C, H, W)
+            input_mask_image = mask_image
+            h, w = image.shape[-2]//self.vae_scale_factor, image.shape[-1]//self.vae_scale_factor
+            mask_image = torchvision.transforms.functional.resize(mask_image, (h,w))
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:532]: Can't use `torchvision`

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py [hunk 3]
```diff
@@ -44,17 +45,26 @@ def preprocess_image(image):
     return 2.0 * image - 1.0
 
 
+# add additional output for mask tensor corresponding to input image (in addition to latent mask)
 def preprocess_mask(mask, scale_factor=8):
     mask = mask.convert("L")
     w, h = mask.size
     w, h = map(lambda x: x - x % 32, (w, h))  # resize to integer multiple of 32
+
+    input_mask = np.array(mask).astype(np.float32) / 255.0
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:54]: This looks a bit backwards breaking - what are we doing here exactly?

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:54]: Could you provide a code snippet explaining why this is needed maybe? :-)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py (hunk 1 above)
Comment (1lint) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:24]: Ah okay, resizing can also be done with [torch interpolate](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html). I will resubmit PR with the change

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py (hunk 3 above)
Comment (1lint) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:54]: The input_mask keeps track of the drawn mask, so that after decoding the inpainted image, we can replace the non-masked parts of the output image with the original image, and only replace the masked portions of the output image with the generated inpainted image. 
It's not absolutely necessary, it just addresses an issue I've seen especially noticeable with anime style sd checkpoints, where changing a small part of the latents (corresponding to the drawn mask) causes distortion of other parts of the decoded image, even far from the drawn image, due to the global nature of the VAE decoder.

Comment (1lint) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:54]: I pasted example differences below showing the effect of preserving the original image (first screenshot, colors match) vs using the full inpainted image (colors slightly grayed). On second thought the distortion away from masked area may largely be due to the VAE reconstruction loss. The webui code for the examples is [here](https://github.com/1lint/sdpipe_webui)

preserve_nonmasked_original
entire_inpainted_image
